openlineage, docs: tips for OpenLineage method implementation (#31817)
mistercrunch/Airflow: docs/apache-airflow-providers-openlineage/guides/developer.rst

Implementing OpenLineage in Operators
-------------------------------------

> OpenLineage defines few methods for implementation for Operators. Those are referred to as OpenLineage methods.

Reviewer: ```suggestion
OpenLineage defines few methods for implementation in Operators. Those are referred to as OpenLineage methods.
```

Reviewer 2: It should be _a few_.

https://dictionary.cambridge.org/grammar/british-grammar/little-a-little-few-a-few

> (A) little and (a) few are quantifiers meaning ‘some’. Little and few have negative meanings. We use them to mean ‘not as much as may be expected or wished for’.

Author: Thanks, fixed both mistakes.

      run_facets: dict[str, BaseFacet] = Factory(dict)
      job_facets: dict[str, BaseFacet] = Factory(dict)

OpenLineage integration itself takes care to enrich it with things like general Airflow facets, proper event time and type, creating proper OpenLineage RunEvent.

How to properly implement OpenLineage methods?
==============================================

> There are several things worth noting when implementing OpenLineage in operators.

Reviewer 2: ```suggestion
There are a couple of things worth noting when implementing OpenLineage in operators.
```

I was anticipating a third thing when reading this, but there seems to be only two?

Author: Hmm, I was referring to everything below that line. I guess that's unclear, tried to fix this.

First, do not import OpenLineage methods on top-level, but in OL method itself.
This allows users to use your provider even if they do not have OpenLineage provider installed.

> 

Reviewer: can we add an example here (or) link to an already existing example? The example will be useful for folks who are starting out on a new OL implementation

Author: Added example from `GcsToGcsOperator` - due to it's simplicity.

Another important point is to make sure your provider returns OpenLineage-compliant dataset names.
It allows OpenLineage consumers to properly match information about datasets gathered from different sources.

Naming convention is described in the `OpenLineage docs <https://openlineage.io/docs/spec/naming>`__.

OpenLineage implementation should not waste time of users that do not use it.
This means not doing heavy processing or network calls in the ``execute`` method that aren't used by it.

> Better option is to save relevant information in Operator attribute - and then use it

Reviewer 3: ```suggestion
Better option is to save relevant information in Operator attributes - and then use it
```

Author: Thanks, fixed.

in OpenLineage method.

> Good example is ``BigQueryExecuteQueryOperator``. It saves ``job_ids`` of queries that were executer.

Reviewer 3: ```suggestion
Good example is ``BigQueryExecuteQueryOperator``. It saves ``job_ids`` of queries that were executed.
```

Author: Thanks, fixed.
